tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+extern crate futures;
+extern crate hdrhistogram;
+#[macro_use]
+extern crate log;
+extern crate tokio_timer;
+extern crate tower_filter;
+extern crate tower_service;
+
+use futures::future;
+use futures::future::FutureResult;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tower_filter::Filter;
+
+mod delay;
+mod latency;
+mod rotating_histogram;
+mod select;
+
+use delay::Delay;
+use latency::Latency;
+use rotating_histogram::RotatingHistogram;
+use select::Select;
+
+type Histo = Arc<Mutex<RotatingHistogram>>;
+pub type Service<InnerSvc, P> = select::Select<
+    SelectPolicy<P>,
+    Latency<Histo, InnerSvc>,
+    Delay<DelayPolicy, Filter<Latency<Histo, InnerSvc>, PolicyPredicate<P>>>,
+>;
+type Error = Box<dyn std::error::Error + Send + Sync>;
+
+/// A policy which describes which requests can be cloned and then whether those
+/// requests should be retried.
+pub trait Policy<Request> {
+    /// clone_request is called when the request is first received to determine
+    /// if the request is retryable.
+    fn clone_request(&self, req: &Request) -> Option<Request>;
+    /// can_retry is called after the hedge timeout to determine if the hedge
+    /// retry should be issued.
+    fn can_retry(&self, req: &Request) -> bool;
+}
+
+#[derive(Clone)]
+pub struct PolicyPredicate<P>(P);
```

Comment:
Are these types supposed to be part of the public API? Or are they only `pub` because they needed to be to satisfy `pub type Service = ..`? If the latter, we could make these all private by change `type Service` into an opaque wrapper `struct` instead.

Reply:
Good call, thanks.  I've added opaque wrappers for the Service and Future types.